src/clj_pgp/message.clj

```diff
@@ -241,70 +254,82 @@
   ;; Read through the list of encrypted session keys and attempt to find one
   ;; which the decryptor will unlock. If none are found, the message is not
   ;; decipherable and an exception is thrown.
-  (unpack-message
+
+  (reduce-message
+    [packet acc r-fn opts]
+    (when-not (can-decrypt? packet opts)
+      (throw (IllegalArgumentException.
+               (str "Cannot decrypt " (pr-str packet) " with " (pr-str opts)
+                    " (no matching encrypted session key)"))))
+    (reduce-message
+      (->> (.getEncryptedDataObjects packet)
+           iterator-seq
+           (filter #(can-decrypt? % opts))
```

Comment:
How about we call that function `readable-object` and it returns the first possible readable object or nil if it can't find one.  The rest of the implementations will return nil or themselves.